parcels/kernelsoa.py

```diff
@@ -0,0 +1,370 @@
+        # Clean-up the in-memory dynamic linked libraries.
+        # This is not really necessary, as these programs are not that large, but with the new random
+        # naming scheme which is required on Windows OS'es to deal with updates to a Parcels' kernel.)
+        super(KernelSOA, self).__del__()
+
+    def __add__(self, kernel):
+        if not isinstance(kernel, KernelSOA):
+            kernel = KernelSOA(self.fieldset, self.ptype, pyfunc=kernel)
+        return self.merge(kernel, KernelSOA)
+
+    def __radd__(self, kernel):
+        if not isinstance(kernel, KernelSOA):
+            kernel = KernelSOA(self.fieldset, self.ptype, pyfunc=kernel)
+        return kernel.merge(self, KernelSOA)
+
+    def remove_deleted(self, pset, output_file, endtime):
+        """
+        Utility to remove all particles that signalled deletion
+
+        This deletion function is targetted to index-addressable, random-access array-collections.
+        """
+        # Indices marked for deletion.
+        bool_indices = pset.collection.state == OperationCode.Delete
+        indices = np.where(bool_indices)[0]
+        if len(indices) > 0 and output_file is not None:
+            output_file.write(pset, endtime, deleted_only=bool_indices)
+        pset.remove_indices(indices)
+
+    def execute(self, pset, endtime, dt, recovery=None, output_file=None, execute_once=False):
+        """Execute this Kernel over a ParticleSet for several timesteps"""
+        pset.collection.state[:] = StateCode.Evaluate
+
+        if abs(dt) < 1e-6 and not execute_once:
+            logger.warning_once("'dt' is too small, causing numerical accuracy limit problems. Please chose a higher 'dt' and rather scale the 'time' axis of the field accordingly. (related issue #762)")
+
+        if recovery is None:
+            recovery = {}
+        elif ErrorCode.ErrorOutOfBounds in recovery and ErrorCode.ErrorThroughSurface not in recovery:
+            recovery[ErrorCode.ErrorThroughSurface] = recovery[ErrorCode.ErrorOutOfBounds]
+        recovery_map = recovery_base_map.copy()
+        recovery_map.update(recovery)
+
+        if pset.fieldset is not None:
+            for g in pset.fieldset.gridset.grids:
+                if len(g.load_chunk) > g.chunk_not_loaded:  # not the case if a field in not called in the kernel
+                    g.load_chunk = np.where(g.load_chunk == g.chunk_loaded_touched,
+                                            g.chunk_deprecated, g.load_chunk)
+
+        # Execute the kernel over the particle set
+        if self.ptype.uses_jit:
+            self.execute_jit(pset, endtime, dt)
+        else:
+            self.execute_python(pset, endtime, dt)
+
+        # Remove all particles that signalled deletion
+        self.remove_deleted(pset, output_file=output_file, endtime=endtime)   # Generalizable version!
+
+        # Identify particles that threw errors
+        n_error = pset.num_error_particles
+
+        # while np.any(error_particles):
```

Comment:
Again a stray debug comment?

Reply:
kind-of - and kind-of not, because the `np.any(...)` function is faster than the `n_error` count. The count for the is (or: was) useful in making the `execute(...)` function work when we implemented the new ParticleSet class structure, which is why we left it in.

So: this comment is fast, and it is a feasible alternative to the `n_error` count, but the `n_error` count is easier when tracing errors for the kernel.

I can remove it - but I need to make a note about it for myself when for implementing all the new kernels later.

Reply:
OK, no worries. Then leave it in